File: examples/cpp/RVC2/ToF/tof_minimal.cpp
#include <iostream>
#include <opencv2/opencv.hpp>
#include <xtensor/containers/xadapt.hpp>
#include <xtensor/containers/xarray.hpp>

#include "depthai/depthai.hpp"

cv::Mat colorizeDepth(const cv::Mat& frameDepth) {
    // -----------------------------------------------------------------------
    // 1.  Basic checks & convert to CV_32F
    // -----------------------------------------------------------------------
    if(frameDepth.empty() || frameDepth.channels() != 1) return cv::Mat::zeros(frameDepth.size(), CV_8UC3);

    cv::Mat depth32f;
    frameDepth.convertTo(depth32f, CV_32F);  // safe for any input type

    // -----------------------------------------------------------------------
    // 2.  Build mask of valid (non-zero) pixels
    // -----------------------------------------------------------------------
    const cv::Mat nonZeroMask = depth32f != 0.0f;
    const int nz = cv::countNonZero(nonZeroMask);
    if(nz == 0) return cv::Mat::zeros(frameDepth.size(), CV_8UC3);

    // -----------------------------------------------------------------------
    // 3.  3 % / 95 % percentiles (identical to Python version)
    // -----------------------------------------------------------------------
    std::vector<float> values;
    values.reserve(nz);
    for(int r = 0; r < depth32f.rows; ++r) {
        const float* d = depth32f.ptr<float>(r);
        const uchar* m = nonZeroMask.ptr<uchar>(r);
        for(int c = 0; c < depth32f.cols; ++c)
            if(m[c]) values.push_back(d[c]);
    }

    std::sort(values.begin(), values.end());
    auto pct = [&](double p) {
        size_t idx = static_cast<size_t>(std::round((p / 100.0) * (values.size() - 1)));
        return values[idx];
    };

    const float minDepth = pct(3.0);
    const float maxDepth = pct(95.0);

    // -----------------------------------------------------------------------
    // 4.  Logarithm (zeros replaced by minDepth to avoid -inf)
    // -----------------------------------------------------------------------
    cv::Mat logDepth;
    depth32f.copyTo(logDepth);
    logDepth.setTo(minDepth, ~nonZeroMask);  // overwrite zeros
    cv::log(logDepth, logDepth);

    const float logMinDepth = std::log(minDepth);
    const float logMaxDepth = std::log(maxDepth);

    // -----------------------------------------------------------------------
    // 5.  Clip & linearly scale to [0,255]  (same as np.interp)
    // -----------------------------------------------------------------------
    cv::min(logDepth, logMaxDepth, logDepth);
    cv::max(logDepth, logMinDepth, logDepth);
    logDepth = (logDepth - logMinDepth) * (255.0f / (logMaxDepth - logMinDepth));

    cv::Mat depth8U;
    logDepth.convertTo(depth8U, CV_8U);

    // -----------------------------------------------------------------------
    // 6.  Colour map + set invalid pixels to black
    // -----------------------------------------------------------------------
    cv::Mat depthFrameColor;
    cv::applyColorMap(depth8U, depthFrameColor, cv::COLORMAP_JET);
    depthFrameColor.setTo(cv::Scalar::all(0), ~nonZeroMask);

    return depthFrameColor;
}

int main() {
    dai::Pipeline pipeline;

    // ToF node
    dai::CameraBoardSocket socket = dai::CameraBoardSocket::AUTO;
    dai::ImageFiltersPresetMode presetMode = dai::ImageFiltersPresetMode::TOF_MID_RANGE;
    std::shared_ptr<dai::node::ToF> tof = pipeline.create<dai::node::ToF>()->build(socket, presetMode);

    // Output queues
    std::shared_ptr<dai::MessageQueue> depthQueue = tof->depth.createOutputQueue();
    std::shared_ptr<dai::MessageQueue> depthRawQueue = tof->rawDepth.createOutputQueue();

    pipeline.start();
    while(pipeline.isRunning()) {
        // Visualize raw depth (unfiltered depth directly from the ToF sensor)
        std::shared_ptr<dai::ImgFrame> depthRaw = depthRawQueue->get<dai::ImgFrame>();
        cv::Mat depthRawImage = colorizeDepth(depthRaw->getCvFrame());
        cv::imshow("depthRaw", depthRawImage);

        // Visualize depth (which is filtered depthRaw)
        std::shared_ptr<dai::ImgFrame> depth = depthQueue->get<dai::ImgFrame>();
        cv::Mat depthImage = colorizeDepth(depth->getCvFrame());
        cv::imshow("depth", depthImage);

        if(cv::waitKey(1) == 'q') break;
    }

    return 0;
}

File: examples/python/RVC2/ToF/tof_minimal.py
#!/usr/bin/env python3
"""Minimal RVC2 ToF example — host ImageFilters depth + raw depth from ToFBase."""

import cv2
import depthai as dai
import numpy as np


def colorize_depth(frame_depth: np.ndarray) -> np.ndarray:
    invalid_mask = frame_depth == 0
    try:
        min_depth = np.percentile(frame_depth[frame_depth != 0], 3)
        max_depth = np.percentile(frame_depth[frame_depth != 0], 95)
        log_depth = np.log(frame_depth, where=frame_depth != 0)
        log_min_depth = np.log(min_depth)
        log_max_depth = np.log(max_depth)
        if log_max_depth <= log_min_depth:
            # Valid depths collapsed to a single value; avoid divide-by-zero normalization.
            log_max_depth = log_min_depth + 1.0
        np.nan_to_num(log_depth, copy=False, nan=log_min_depth)
        log_depth = np.clip(log_depth, log_min_depth, log_max_depth)
        depth_color = np.interp(log_depth, (log_min_depth, log_max_depth), (0, 255))
        depth_color = depth_color.astype(np.uint8)
        depth_color = cv2.applyColorMap(depth_color, cv2.COLORMAP_JET)
        depth_color[invalid_mask] = 0
        return depth_color
    except IndexError:
        return np.zeros((frame_depth.shape[0], frame_depth.shape[1], 3), dtype=np.uint8)


def main() -> None:
    pipeline = dai.Pipeline()

    tof = pipeline.create(dai.node.ToF).build(
        dai.CameraBoardSocket.AUTO,
        dai.ImageFiltersPresetMode.TOF_MID_RANGE,
    )

    depth_q = tof.depth.createOutputQueue()
    raw_depth_q = tof.rawDepth.createOutputQueue()

    with pipeline as p:
        p.start()
        while p.isRunning():
            raw = raw_depth_q.tryGet()
            if raw is not None:
                cv2.imshow("rawDepth (ToFBase)", colorize_depth(raw.getFrame()))

            depth = depth_q.tryGet()
            if depth is not None:
                cv2.imshow("depth (ImageFilters)", colorize_depth(depth.getFrame()))

            if cv2.waitKey(1) == ord("q"):
                break


if __name__ == "__main__":
    main()
